### PR TITLE
Dev mega refresh token

### PR DESCRIPTION
**Pull request by weijjia**
Changes:
1. modify cache logic to support cross tenant refresh token
2. modify the cache logic for current device code flow, acquireTokenWithDeviceCode won't do cache lookup, it will only store returned token into cache, for cache lookup developer needs to call acquireToken. 
3. add another property 'extraQP' in authenticationcontext. This could be used to set the extra query parameter passed to tokenendpoint.

### CODE REVIEW

File: lib/token-request.js [hunk 1]
```diff
@@ -163,6 +163,22 @@ TokenRequest.prototype._getTokenWithCacheWrapper = function(callback, getTokenFu
   });
 };
 
+TokenRequest.prototype._storeTokenWithCacheWrapper = function(callback, getTokenFunc) {
```
Comment (RandalliLama) [inline lib/token-request.js]: I think this function is more complicated than it needs to be.  I like factoring out the code that adds the token to the cache, however, in this case you do not need to include the getToken call in the same function that adds the token to the cache.  Leave that bit of code in getTokenWithDeviceCode.  When the getToken function returns, or calls the callback, then call a separate function to add the token to the cache.  Something like:

``` js
 this._oauthGetTokenByPolling(oauthParameters, interval, expires_in, function(err) {
   if (err) {
  } else {
   addTokenToCache(callback)
  } );
```

File: sample/device-code-sample.js [hunk 2]
```diff
@@ -93,8 +97,20 @@ context.acquireUserCode(resource, sampleParameters.clientId, 'es-mx', function (
                 console.log(err);
             }
             else {
-                console.log(tokenResponse);
+                authorityUrl = sampleParameters.authorityHostUrl + '/' + sampleParameters.anothertenant;
+                
+                var context2 = new AuthenticationContext(authorityUrl, null, cache);
```
Comment (RandalliLama) [inline sample/device-code-sample.js]: Why are you adding a second AuthenticationContext?

File: test/cache-driver.js [hunk 3]
```diff
@@ -35,7 +35,7 @@ var testRequire = util.testRequire;
 var assert = require('assert');
 
 var MemoryCache = testRequire('memory-cache');
-var CacheDriver = testRequire('cache-driver');
+var CacheDriver = testRequire('../lib/cache-driver');
```
Comment (RandalliLama) [inline test/cache-driver.js]: Why is this necessary.  The whole point of the testRequire function was to avoid having to do this. :)

File: test/cache-driver.js [hunk 4]
```diff
@@ -231,7 +232,7 @@ suite('CacheDriver', function() {
 
         var otherAuthority = 'someOtherAuthority';
         var responseOptions = { authority : otherAuthority, mrrt : true, resource : responses[0].resource };
-        var differentAuthorityResponse = util.createResponse(responseOptions);
+        var differentAuthorityResponse = util.createResponse(responseOptions, 21);
```
Comment (RandalliLama) [inline test/cache-driver.js]: Magic number?  Why 21?  Also, could you add a comment to the definition of the util.createResponse function that explains what the iteration parameter was for.  It took a while to refresh my memory.  I should have doc'ed it in the first place.  Something like "iteration - createResponse will create a distinct token for each value of iteration and it will always return that same token for the same value of iteration.  This is useful when you need a non singleton set of tokens to test a cache scenario against."

**Comment by RandalliLama** [comment]
I'm worried about CacheDriver._updateRefreshTokens.  What about the case that the client is interacting with a server that has tenant specific refresh tokens?  How will we know whether to update all of the refresh tokens regardless of tenant vs. just the ones for a particular tenant?

File: lib/authentication-context.js [hunk 5]
```diff
@@ -142,6 +142,12 @@ Object.defineProperty(AuthenticationContext.prototype, 'options', {
   }
 });
 
+Object.defineProperty(AuthenticationContext.prototype, 'extraQP', {
```
Comment (RandalliLama) [inline lib/authentication-context.js]: Please remove this.  It's not related to the current feature and shouldn't go in here.  We need to be *really* careful about how we change the public API.  No changes should happen to the Api without team discussion.  We need to ensure that the API's are largely the same across platforms.

File: sample/device-code-sample.js (hunk 2 above)
Comment (weijjia) [inline sample/device-code-sample.js]: this is for the cross tenant refresh token testing, i want to show the way refresh tokens works for cross tenant refresh token.

File: lib/oauth2client.js [hunk 6]
```diff
@@ -87,19 +87,31 @@ OAuth2Client.prototype._createTokenUrl = function () {
   var tokenUrl = url.parse(this._tokenEndpoint);
 
   var parameters = {};
-  parameters.slice = 'testslice';
   parameters[OAuth2Parameters.AAD_API_VERSION] = '1.0';
 
-  tokenUrl.search = querystring.stringify(parameters);
+  var queryParameter = querystring.stringify(parameters);
+  var extraQueryParam = this._callContext.extraQP;
```
Comment (RandalliLama) [inline lib/oauth2client.js]: Same as in authentication-context.js.  Please remove.

File: test/cache-driver.js (hunk 3 above)
Comment (weijjia) [inline test/cache-driver.js]: ah, my bad, I should remove it.

File: lib/oauth2client.js [hunk 7]
```diff
@@ -433,7 +445,7 @@ OAuth2Client.prototype._createPostOption = function (postUrl, urlEncodedRequestF
     var postOptions = util.createRequestOptions(
         this,
     {
-            'url' : postUrl,
+            'url' : url.format(postUrl),
```
Comment (RandalliLama) [inline lib/oauth2client.js]: Was there a bug here?

File: lib/cache-driver.js [hunk 8]
```diff
@@ -137,10 +137,6 @@ function CacheDriver(callContext, authority, resource, clientId, cache, refreshF
  * @param  {QueryCallback} callback
  */
 CacheDriver.prototype._find = function(query, callback) {
-  var authorityQuery = {};
```
Comment (RandalliLama) [inline lib/cache-driver.js]: I thought that the behavior we are after is to always look for a tenant specific token first.  If we can't find one of those then we attempt to us an MRRT to get a token for the other tenant.  If that doesn't work the we just fail.  I don't see that happening here.  You've removed the authority clause of the cache queries, but I don't see where it's been added back anywhere in more particular circumstances.  Did I miss it?

File: lib/authentication-context.js (hunk 5 above)
Comment (weijjia) [inline lib/authentication-context.js]: Sure, it makes sense, and will remove it. I saw in the android-adal and adal-js, they all have ways to specify extra query parameter sent to request. This could kinda avoid the slice=testslice mistake we saw in the master branch.

File: test/cache-driver.js (hunk 4 above)
Comment (weijjia) [inline test/cache-driver.js]: :) it's 21 because of the specific test case, we add 20 entries into cache, and then add another 2 entries into cache with isMRRT=true. How it used to work is, one doesn't have the iteration number when create the response, and the other one has 21(this is simply there is already 20 entries, and then 21 will be the newly added one). This works totally fine before since we look up cache by userid, clientid, and authority, however with the cross tenant refresh token, we will update the refresh token of the cached entry if there is a match of user id, and client id. That's why I change it to 21, just to make it the same with the entry added after it. 

I agree, will add the doc.

File: sample/device-code-sample.js (hunk 2 above)
Comment (RandalliLama) [inline sample/device-code-sample.js]: Got it.  Makes sense.  I missed that.

File: lib/authentication-context.js (hunk 5 above)
Comment (RandalliLama) [inline lib/authentication-context.js]: The extra query parameters you see in the other platforms only send it to the authorize endpoint as far as I know.  No platform sends it to the token endpoint.

File: lib/oauth2client.js (hunk 7 above)
Comment (weijjia) [inline lib/oauth2client.js]: I think it's a bug. We never send down any QP in the post request.

File: lib/cache-driver.js (hunk 8 above)
Comment (weijjia) [inline lib/cache-driver.js]: We firstly find all potential entries, and then find tentant and resource specific entry. 
check here https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/pull/83/files#diff-8219131514daab4a7f8d1b3f58959918R204

File: lib/oauth2client.js (hunk 7 above)
Comment (weijjia) [inline lib/oauth2client.js]: but if no platform actually send down any qp to token endpoint, then we should remove the api-version=1.0, it doesn't need to be there.

**Comment by kpanwar** [comment]
The release difference between other instances and PROD is at worst case a week of difference. We can release the library all other instances are also updated with the feature changes instead of asking server to add a flag in response for such a short period of time,

File: lib/oauth2client.js (hunk 7 above)
Comment (RandalliLama) [inline lib/oauth2client.js]: Before fixing this I'd like to see if we can determine why it was not being sent.  I am almost positive that at one point it was definitely being sent.  If it is not being sent then we should fix it so that it is sent.

File: lib/cache-driver.js (hunk 8 above)
Comment (RandalliLama) [inline lib/cache-driver.js]: ok.

**Comment by RandalliLama** [comment]
ok to only releasing new package after all AAD Prod stamps are updated.

**Comment by RandalliLama** [comment]
:shipit:
